bootstrap/main.py

```diff
@@ -0,0 +1,194 @@
+    sentinel = join(directory, WEST_HIDDEN_SENTINEL)
+    os.mkdir(sentinel)
+    # On Windows, we have to manually mark a file as hidden.
+    if platform.system() == 'Windows':
+        subprocess.check_call(['attrib', '+H', sentinel])
+
+
+def clone(url, rev, dest):
+    def repository_type(url):
+        if (url.startswith('http:') or url.startswith('https:') or
+                url.startswith('git:') or url.startswith('git+ssh:') or
+                url.startswith('file:')):
+            return 'GIT'
+        else:
+            return 'UNKNOWN'
+
+    if os.path.exists(dest):
+        msg = 'refusing to clone into existing location {}'.format(dest)
+        raise WestError(msg)
+
+    rtype = repository_type(url)
+    if rtype == 'GIT':
+        subprocess.check_call(['git', 'clone', url, '-b', rev, dest])
+    else:
+        sys.exit('Unknown repository URL scheme for URL {}'.format(url))
+
+
+#
+# west init
+#
+
+
+def init(argv):
+    '''Command line handler for ``west init`` invocations.
+
+    This exits the program with a nonzero exit code if fatal errors occur.'''
+    init_parser = argparse.ArgumentParser(
+        prog='west init',
+        description='Bootstrap initialize a Zephyr installation')
+    init_parser.add_argument(
+        '-u', '--manifest-url', default=MANIFEST_DEFAULT,
+        help='Zephyr manifest fetch URL, default ' + MANIFEST_DEFAULT)
+    init_parser.add_argument(
+        '--mr', '--manifest-rev', default=MANIFEST_REV_DEFAULT,
+        dest='manifest_rev',
+        help='Manifest revision to fetch, default ' + MANIFEST_REV_DEFAULT)
+    init_parser.add_argument(
+        '-w', '--west-url', default=WEST_DEFAULT,
+        help='West fetch URL, default ' + WEST_DEFAULT)
+    init_parser.add_argument(
+        '--wr', '--west-rev', default=WEST_REV_DEFAULT, dest='west_rev',
+        help='West revision to fetch, default ' + WEST_REV_DEFAULT)
+    init_parser.add_argument(
+        'directory', nargs='?', default=None,
+        help='Initializes in this directory, creating it if necessary')
+
+    args = init_parser.parse_args(args=argv)
+    directory = args.directory or os.getcwd()
+
+    try:
+        topdir = find_west_topdir(directory)
+        init_reinit(topdir, args)
+    except WestNotFound:
+        init_bootstrap(directory, args)
+
+
+def init_bootstrap(directory, args):
+    '''Bootstrap a new manifest + West installation in the given directory.'''
+    if not os.path.isdir(directory):
+        try:
+            os.makedirs(directory, exist_ok=False)
+        except PermissionError:
+            sys.exit('Cannot initialize in {}: permission denied'.format(
+                directory))
+        except FileExistsError:
+            sys.exit('Something else created {} concurrently; quitting'.format(
+                directory))
+        except Exception as e:
+            sys.exit("Can't create directory {}: {}".format(
+                directory, e.args))
+
+    make_topdir_sentinel(directory)
```

Comment:
Unwinding this sequence on error would need more careful handling if we want to get worried about it failing. I'm hoping to keep it simple here, though, since the whole idea is that the bootstrap directory isn't in use by anything else.

Reply:
Should we clone west into the sentinel folder? I don't think so but just checking